feat!: add input mr into genesis block

hansieodendaal opened the pull request:
Description
---
- Added the input merkle root calculation into the genesis block; previously, this was just a default hash, which is incorrect. This will enable adding immediate spending of pre-mine outputs in the genesis block.
- Added total kernel offset to the pre-mine output immediate spend ceremony.
- Added total kernel offset to the pre-mine information added to the genesis block.
- Added chain balance for immediate spend inputs and outputs in `xxxx_genesis_sanity_check()` unit tests.

Motivation and Context
---
The genesis block did not cater for any immediate spend inputs.

How Has This Been Tested?
---
- Unit tests expanded & pass.
- System-level tests:
  - Performed a new pre-mine creation ceremony (_esmeralda_).
  - Performed a pre-mine immediate spend ceremony.
  - Added the pre-mine and pre-mine immediate spend to the _esmeralda_pre_mine,json_ embedded input file.
  - Unit test `fn esmeralda_genesis_sanity_check()` passes.

What process can a PR reviewer use to test or verify this change?
---
Code review.

<!-- Checklist -->
<!-- 1. Is the title of your PR in the form that would make nice release notes? The title, excluding the conventional commit
tag, will be included exactly as is in the CHANGELOG, so please think about it carefully. -->

Breaking Changes
---

- [ ] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [X] Other - Please specify

<!-- Does this include a breaking change? If so, include this line as a footer -->
**BREAKING CHANGE:** The genesis block changed; this requires a blockchain restart.

SWvheerden commented on base_layer/core/src/blocks/genesis_block.rs:
```diff
@@ -180,6 +201,8 @@ pub fn get_nextnet_genesis_block() -> ChainBlock {
         block.header.kernel_mr =
             FixedHash::from_hex("36881d87e25183f5189d2dca5f7da450c399e7006dafd9bd9240f73a5fb3f0ad").unwrap();
+        block.header.input_mr =
+            FixedHash::from_hex("212ce6f5f7fc67dcb73b2a8a7a11404703aca210a7c75de9e50d914c9f9942c2").unwrap();
```
```suggestion
            FixedHash::zero();
```
So that it does not reset the network this time around

SWvheerden reviewed (commented):
This looks good, just dont make it a reset now for nextnet

Commit:
Add input mr into genesis block

Added the input merkle root calculation into the genesis block; previously
this was just a default hash, which is incorrect. This will enable adding
immediate spending of pre-mine outputs in the genesis block.

SWvheerden commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -247,16 +250,50 @@ where B: BlockchainBackend
             blockchain_db.write(txn)?;
             blockchain_db.store_pruning_horizon(config.pruning_horizon)?;
+            // Sanity check
+            match blockchain_db.fetch_block_by_hash(*genesis_block.hash(), false)? {
+                Some(block) => {
+                    let header_from_db = block.block().clone().header;
+                    let header_from_code = genesis_block.block().clone().header;
+                    if header_from_db != header_from_code {
+                        return Err(ChainStorageError::CriticalError(
+                            "Genesis block header  in db does not match genesis block in code".into(),
+                        ));
+                    }
+                    let mut body_from_db = block.block().clone().body;
+                    body_from_db.sort();
+                    let mut body_from_code = genesis_block.block().clone().body;
+                    body_from_code.sort();
+                    if body_from_db != body_from_code {
+                        return Err(ChainStorageError::CriticalError(
+                            "Genesis block body in db does not match genesis block in code".into(),
+                        ));
+                    }
+                },
+                None => {
+                    return Err(ChainStorageError::CriticalError(
+                        "Genesis block could not be created".into(),
+                    ))
+                },
+            }
```
Why add this, you are checking if what you just added is in the database?
If that would have failed, it would have thrown an error in ln 269
All this is doing is adding more lines. This code should assume that the lmdb code is working. We should assume that lmdb has its own unit tests to ensure if a write fails, it will error. Or that what you write is actually in the database

Commit:
review

Commit:
total kernel offset
